app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -86,42 +84,110 @@ async function addEthereumChainHandler(
+
+      const clonedNetwork = { ...existingNetwork };
+
+      // Check if the rpc url already exists
+      let rpcIndex = clonedNetwork.rpcEndpoints.findIndex(
+        ({ url }) => url === firstValidRPCUrl,
+      );
+
+      // If it doesn't exist, add a new one
+      if (rpcIndex === -1) {
+        clonedNetwork.rpcEndpoints = [
+          ...clonedNetwork.rpcEndpoints,
+          {
+            url: firstValidRPCUrl,
+            type: RpcEndpointType.Custom,
+            name: chainName,
+          },
+        ];
+        rpcIndex = clonedNetwork.rpcEndpoints.length - 1;
+      }
+
+      // The provided rpc endpoint becomes the default
+      clonedNetwork.defaultRpcEndpointIndex = rpcIndex;
 
-      networkClientId = await upsertNetworkConfiguration(
-        {
-          chainId,
-          rpcPrefs: { blockExplorerUrl: firstValidBlockExplorerUrl },
-          nickname: chainName,
-          rpcUrl: firstValidRPCUrl,
-          ticker,
-        },
-        { source: 'dapp', referrer: origin },
+      // Check if the block explorer already exists
+      let blockExplorerIndex = clonedNetwork.blockExplorerUrls.findIndex(
+        (url) => url === firstValidBlockExplorerUrl,
       );
-    } catch (error) {
-      endApprovalFlow({ id: approvalFlowId });
-      return end(error);
+
+      // If it doesn't exist, add a new one
+      if (blockExplorerIndex === -1) {
+        clonedNetwork.blockExplorerUrls = [
+          ...clonedNetwork.blockExplorerUrls,
+          firstValidBlockExplorerUrl,
+        ];
+        blockExplorerIndex = clonedNetwork.blockExplorerUrls.length - 1;
+      }
+
+      // The provided block explorer becomes the default
+      clonedNetwork.defaultBlockExplorerUrlIndex = blockExplorerIndex;
+
+      updatedNetwork = await updateNetwork(
+        clonedNetwork.chainId,
+        clonedNetwork,
+        currentChainId === chainId
+          ? {
+              replacementSelectedRpcEndpointIndex:
+                clonedNetwork.defaultRpcEndpointIndex,
+            }
+          : undefined,
+      );
+    } else {
+      // A network for this chain id does not exist, so add a new network
+      updatedNetwork = await addNetwork({
+        blockExplorerUrls: firstValidBlockExplorerUrl
+          ? [firstValidBlockExplorerUrl]
+          : [],
+        defaultBlockExplorerUrlIndex: firstValidBlockExplorerUrl
+          ? 0
+          : undefined,
+        chainId,
+        defaultRpcEndpointIndex: 0,
+        name: chainName,
+        nativeCurrency: ticker,
+        rpcEndpoints: [
+          {
+            url: firstValidRPCUrl,
+            name: chainName,
+            type: RpcEndpointType.Custom,
+          },
+        ],
+      });
     }
+  } catch (error) {
+    endApprovalFlow({ id: approvalFlowId });
+    return end(error);
+  }
 
-    requestData = {
+  // If the added or updated network is not the current chain, prompt the user to switch
+  if (chainId !== currentChainId) {
```

Comment:
It's possible this should compare against `currentChainIdForDomain` instead of `currentChainId`.  But I'm not sure in what cases they can be different.

It may also affect when we pass `replacementSelectedRpcEndpointIndex` to `updateNetwork` above, which must be done when editing the current network.

Reply:
https://github.com/MetaMask/metamask-extension/pull/26433/commits/c46638ac0898ffca7622c16bc39d0b2381eb5eee